TST tight tests for GLMs

lorentzenchr opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
This is similar to #22910, but for GLMs instead of Ridge.

#### What does this implement/fix? Explain your changes.
Very tight test suite for penalized and unpenalized GLMs.

It also **modifies the lbfgs options** for better convergence (such that it passes the new tests). I think this does not need a changelog entry, but it modifies the model (it may need more iterations, but produce more precise coefficients).


#### Any other comments?
Prerequesite for #23314.

Locally, all tests pass with `SKLEARN_TESTS_GLOBAL_RANDOM_SEED="all" pytest -x -Werror sklearn/linear_model/_glm/tests/test_glm.py`.

lorentzenchr committed:
TST add rigoros test for GLMs

lorentzenchr committed:
TST improve test_warm_start

lorentzenchr committed:
ENH improve lbfgs options for better convergence

lorentzenchr committed:
CLN fix test_warm_start

ogrisel commented:
I think we should document the change in the internal LBFGS configration to document the improved convergence and maybe suggest the user to adapt the `tol` parameter accordingly if they do not require tight convergence.

ogrisel commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -285,9 +285,10 @@ def fit(self, X, y, sample_weight=None):
                     "iprint": (self.verbose > 0) - 1,
                     "gtol": self.tol,
-                    "ftol": 1e3 * np.finfo(float).eps,
+                    "ftol": 64 * np.finfo(float).eps,  # lbfgs is float64 land.
```
What's the justification for the 64 magic number?

lorentzenchr replied:
Yes, magic. It's just smaller than 1e3 which was also just magic. 64 is a power of 2 so ftol will be a number with all bits equal zero except one single bit.

Finding good stopping criteria is just hard. For high precision (small tol), we sometimes have tiny but still significant loss differences, e.g. <~ 1e-14 but maximum |gradient| still larger than tol.

ogrisel replied:
I agree the previous `ftol` value was probably too lax.

lorentzenchr replied:
As said earlier: 64 is just a smaller number than the previous value 1e3. I tested lower and higher values. This one passes this new test suite.

ogrisel replied:
Since two of us asked the same question maybe it's a sign that this answer should be written as an inline comment (e.g. mentioning that the 64 magic constant was found iteratively to make the test suite pass with the lowest possible tolerance values).

lorentzenchr replied:
Done in https://github.com/scikit-learn/scikit-learn/pull/23619/commits/2fa4397bd198c932a851d63b0feb1313c2264014.

lorentzenchr commented:
> I think we should document the change in the internal LBFGS configuration to document the improved convergence

I'll add a changelog entry.

> maybe suggest the user to adapt the tol parameter accordingly if they do not require tight convergence.

No objections in general. I just don't think it's necessary as our default tol=1e-4 is quite large.

ogrisel commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,474 @@ def regression_data():
+            assert norm_model > norm_solution
+            pytest.xfail(
+                reason=f"GLM with {solver} does not provide the minimum norm solution."
+            )
```
I think this is not really an valid use of `xfail`: as far as I know, lbfgs has no implicit inductive bias to select the minimum normal solution (but I believe SGD has it).

So we could instead remove the xfail statement here and instead check that re-running the same fit with a very small value of `alpha` (e.g. `alpha = 1e-30`) would actually lead to converging close the minimum norm solution of the unregularized problem.

WDYT?

ogrisel replied:
> We could, however, add it on top as additional check.

Yes this is what I meant:

- first check that lbfgs with alpha=0 converges to a non-minimal normal solution (same predictions and loss as the minimum norm solution to the the constructed unpenalized problem of the fixture)
- then refit lbfgs with alpha=eps and checks that it converges to the minimal norm solution of the constructed unpenalized problem.

lorentzenchr replied:
I tried it and it does not work out: Setting tiny alpha with lbfgs does not give the minimum norm solution. I don't know if that is due to the solver or due to GLMs. This is kind of uncharted territory where I do not know any literature.

See https://github.com/scikit-learn/scikit-learn/pull/23619/commits/382d17707457638ee936bf44a9f25bb3e296a71c.

ogrisel replied:
This is mysterious but I guess making this kind of assertions is too dependent on details such as the precise stopping criteria of LBFGS, the condition number of the hessian near in the subspace of solutions for the unpenalized problem.

Thanks for trying.

ogrisel commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,474 @@ def regression_data():
+            assert norm_model > norm_solution
+            pytest.xfail(
+                reason=f"GLM with {solver} does not provide the minimum norm solution."
+            )
```
Same comment as above about the use of xfail.

ogrisel reviewed (commented):
Overall I like the idea and I am ok to change the default parameters of the lbfgs solver to be stricter to allow more precise convergence when setting low `tol` values.

My main concern is the expectations we put behind the use of `xfail` in our tests. I think lbfgs has no reason to converge to the minimum norm solution for unpenalized problems. We should therefore not assume that this is a "bug" of lbfgs. See the inline comments for more details.

ogrisel commented on sklearn/linear_model/_glm/glm.py:
```diff
@@ -285,9 +285,10 @@ def fit(self, X, y, sample_weight=None):
                 options={
                     "maxiter": self.max_iter,
+                    "maxls": 40,  # default is 20
                     "iprint": (self.verbose > 0) - 1,
```
Why not pass `self.verbose - 1` directly to allow printing more detailed information from the lfbgs solver?

Could be tackled in a separate PR if you prefer not to mix concerns.

lorentzenchr replied:
Taken with pleasure.

lorentzenchr committed:
TST fix assert singular values in datasets

lorentzenchr commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,474 @@ def regression_data():
+    X[:, -1] = 1  # last columns acts as intercept
+    U, s, Vt = linalg.svd(X)
+    assert np.all(s) > 1e-3  # to be sure
+    assert np.max(s) / np.min(s) < 100  # condition number
```
I added `condition number of X`. It helps a lot if X is well conditioned. It's a main source of possible ill-conditioning in GLM optimization problems.

lorentzenchr commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,474 @@ def regression_data():
+    assert np.all(s) > 1e-3  # to be sure
+    assert np.max(s) / np.min(s) < 100  # condition number
+    U1, _ = U[:, :k], U[:, k:]
+    Vt1, _ = Vt[:k, :], Vt[k:, :]
```
It stems form copy & pasting from the ridge tests where `U2` is used. Your suggestion is clearly better here.

lorentzenchr commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,474 @@ def regression_data():
+
+    For "wide", we return the minimum norm solution w = X' (XX')^-1 y:
+
+        min ||w||_2 subject to X w = y
```
I corrected the docstring and added a code comment further below.

lorentzenchr commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,474 @@ def regression_data():
+        If "wide", then n_features > n_samples.
+    model : a GLM model
+
+    For "wide", we return the minimum norm solution w = X' (XX')^-1 y:
+
+        min ||w||_2 subject to X w = y
```
Yes, this is the line of reasoning. It has, however, nothing to do with probability distributions. I just the mere fact that the deviance is minimized for y_true = y_predicted = inverse_link(raw_prediction). Therefore, in the wide case, we construct raw_predictions fulfilling this. Finding the corresponding minimal norm coefficients is the a simple OLS that we know how to solve.

Let me know if my new comments are fine.

ogrisel replied:
Yes it's great.

lorentzenchr commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -2,25 +2,60 @@
+        options={"ftol": tol, "xtol": tol, "gtol": tol},
+    )
+    return res.x
+
```
I have no idea 😄 
For high dimensional / feature rich X, Nelder-Mead is pretty bad. But starting with lbfgs and then taking over with `root` could be a funny strategy.

lorentzenchr commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,474 @@ def regression_data():
+
+    We work with a simple constructed data set with known solution.
+    Fit on [X] with alpha is the same as fit on [X, X]/2 with alpha/2.
+    For long X, [X, X] is a singular matrix.
```
[X, X] is still long, but it is a singular matrix.

lorentzenchr committed:
CLN address most review comments

lorentzenchr committed:
ENH enable more vebosity levels for lbfgs

lorentzenchr committed:
DOC add whatsnew

lorentzenchr committed:
Merge branch 'main' into glm_tests

ogrisel commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,479 @@ def regression_data():
+    We work with a simple constructed data set with known solution.
+    Note: This checks the minimum norm solution for wide X, i.e.
+    n_samples < n_features:
+        min ||w||_2 subject to w minimizing the mean deviviance.
```
```suggestion
        min ||w||_2 subject to w minimizing the mean deviance.
```

lorentzenchr committed:
CLN remove xfail and clean a bit

lorentzenchr committed:
CLN docstring about minimum norm

ogrisel committed:
More informative repr for the glm_dataset fixture cases

ogrisel commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,479 @@ def regression_data():
+            norm_model = np.linalg.norm(np.r_[model.intercept_, model.coef_])
+            assert norm_model > (1 + 1e-12) * norm_solution
+
+            # Note: Even adding a tiny penalty does not give the minimal norm solution.
```
Actually, not sure whether this is a bug or not as discussed in https://github.com/scikit-learn/scikit-learn/pull/23619#discussion_r899385081.

ogrisel reviewed (approved):
Assuming CI is green, this LGTM.

Maybe @agramfort would be interested in giving this a second review?

ogrisel committed:
Forgot to run black

agramfort commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,480 @@ def regression_data():
+    """Test that GLM converges for all solvers to correct solution on hstacked data.
+
+    We work with a simple constructed data set with known solution.
+    Fit on [X] with alpha is the same as fit on [X, X]/2 with alpha/2.
```
```suggestion
    Fit on [X] with alpha the same as fit on [X, X]/2 with alpha/2.
```

lorentzenchr replied:
You have to read "A fit .... is the same as a fit on ...". So "fit" is used as noun.

agramfort commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,480 @@ def regression_data():
+            assert norm_model > (1 + 1e-12) * norm_solution
+
+            # Note: Even adding a tiny penalty does not give the minimal norm solution.
+            # model_pen = clone(model).set_params(**params).set_params(alpha=1e-10)
+            # model_pen.fit(X, y)
+            # assert_allclose(model_pen.predict(X), y, rtol=1e-6)  # This is true.
+            # norm_model_pen = np.linalg.norm(
+            #     np.r_[model_pen.intercept_, model_pen.coef_]
+            # )
+            # All the following assertions fail.
+            # assert norm_model_pen == pytest.approx(norm_solution)
+            # assert model_pen.intercept_ == pytest.approx(intercept)
+            # assert_allclose(model_pen.coef_, coef, rtol=rtol)
```
what do we do with this? keep this for later hence adding XXX?
or we cleanup and just remove this?

lorentzenchr replied:
I'll add an XXX as I think having these pieces of code here helps to know exactly what fails.

ogrisel replied:
We could make it a `XXX` as I originally thought in https://github.com/scikit-learn/scikit-learn/pull/23619#pullrequestreview-1009473106 but then I thought I wasn't sure it was a bug or not. I don't know if we can expect that LBFGS should converge to the minimal norm solution in the limit of tiny penalty or how to configure the relative tolerances of the various internal stopping criteria of LBFGS to guarantee this.

That being said, I also don't link commented out code in the code base. Maybe we should instead extract this as a reproducer snippet and put it in an issue instead (and just reference the issue in a comment in the test).

lorentzenchr replied:
I've opend an issue #23670.

agramfort reviewed (commented):
just 2 nitpicks

thx @lorentzenchr

jjerphan commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,480 @@ def regression_data():
+            TweedieRegressor(power=1.5),
+        ],
+    ),
+    ids=lambda param: f"{param[0]}-{param[1]}",
```
Today, I learn this. Thanks!

jjerphan reviewed (commented):
Thank you, @lorentzenchr.

A first pass before getting back on solvers.

lorentzenchr commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -125,7 +125,7 @@ def ols_ridge_dataset(global_random_seed, request):
     X[:, -1] = 1  # last columns acts as intercept
     U, s, Vt = linalg.svd(X)
-    assert np.all(s) > 1e-3  # to be sure
+    assert np.all(s > 1e-3)  # to be sure
```
😄

lorentzenchr commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -2,25 +2,60 @@
+SOLVERS = ["lbfgs"]
+
+
+class BinomialRegressor(_GeneralizedLinearRegressor):
+    def _get_loss(self):
+        return HalfBinomialLoss()
```
Isn't it great: Logistic regression defined in only 3 lines of code!
This way, we can test logistic regression too, with the same solvers. Eventually, we might want to use certain solvers in `LogisticRegression`. This way, they are already tested.

ogrisel replied:
I agree that at some point in the future we will refactor our `LogisticRegression` class to reuse the generic work that was done in `_GeneralizedLinearRegressor` and simplify a lot the code base.

However this is a non trivial refactoring that I would rather not start before we converge on better testing of GLMs with improved solvers such as #23314 and friends.

lorentzenchr replied:
> However this is a non trivial refactoring

What do you suggest/prefer: Remove `BinomialRegressor`? It is 3 lines of code in a test file +1 line to add in the test parametrization. That is hardly a burden.

ogrisel replied:
Side note on the possibility to implement multinomial logistic regression in a similar manner (with a `MultinomialRegressor` class): this could be a little awkward because all regressors in scikit-learn are expected to output a `y_pred` with the same shape as `y_train`, so `y_train` would have to be passed as one-hot codes which is not very memory or computationally efficient to treat multinomial logistic regression as a regressor instead of a classifier with hard predictions in `predict` and soft predictions in `predict_proba`.

Not however that I wouldn't be opposed to include a `MultinomialRegressor` in this test suite but that would require a slightly more complex fixture.

lorentzenchr replied:
Multinomial regression is a much harder problem when one tries to implement it efficiently. The important bits of `LinearModelLoss` are already used in `LogisticRegression` and tested there. It would add extensively more code here.

lorentzenchr commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,480 @@ def regression_data():
+    params = dict(
+        alpha=alpha,
+        fit_intercept=fit_intercept,
+        # solver=solver,  # only lbfgs available
```
It well become handy once there is more than one solver. The parametrization anticipates this.
Are you ok with it or shall I remove it for the time being?

lorentzenchr replied:
As always, it's a bit tricky. For simplicity, I'll simply remove the solver parameter/parameterisation.

`_GeneralizedLinearRegressor` has a `solver` parameter, but the public estimators like `PoissonRegressor` don't. I still prefer to test the public estimators instead of the private one as it is closer to an end-to-end test.

jjerphan replied:
I see, is this proposal sensible regarding what you mentioned ?

```suggestion
        # As `solver` is available for a few estimators and
        # as "lbfgs" is the only solver, we left it commented.
        # TODO: make sure `solver` is exposed for all the
        # estimators class subclassing `_GeneralizedLinearRegressor`
        # and uncomment this line.
        # solver=solver,
```

lorentzenchr replied:
In 28fdbab80f7582510f77a1fb397a28558be8b050 I added
```
        # While _GeneralizedLinearRegressor exposes the solver parameter, public
        # estimators currently do not, and lbfgs is the only solver anyway.
        # TODO: Expose solver as soon as we have a second solver to choose from.
        # solver=solver,  # only lbfgs available
```

lorentzenchr commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -2,25 +2,60 @@
 from numpy.testing import assert_allclose
 import pytest
-import warnings
+from scipy import linalg
```
I removed the LinAlgWarning entirely. They are only needed for certain new solvers of other PRs.

lorentzenchr commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,480 @@ def regression_data():
+        max_iter=1000,
+    )
+
+    if solver == "lbfgs" and not fit_intercept:
+        # Sometimes (depending on global_random_seed) lbfgs fails with:
+        # Line search cannot locate an adequate point after MAXLS function and gradient
+        # evaluations. Previous x, f and g restored.
+        # Possible causes: 1 error in function or gradient evaluation;
+        #                  2 rounding error dominate computation.
+        pytest.xfail()
```
This still fails even with the little change on the lbfgs solver options.
For instance, when I run `SKLEARN_TESTS_GLOBAL_RANDOM_SEED="all" pytest -x -Werror sklearn/linear_model/_glm/tests/test_glm.py`.

lorentzenchr replied:
I consider this test a failure as it does not pass for all random seeds. This way, it is much more transparent that we have an issue here, IMHO.

jjerphan replied:
I see. Personally, I would prefer it to be scoped precisely. What are your preferences, @ogrisel?

ogrisel replied:
I had another look at this and it seems I cannot trigger the warning anymore than that we increased MAXLS again.

I pushed a commit with `[all random seeds]` on this test to check that this is actually the case on all supported platforms.

lorentzenchr commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,480 @@ def regression_data():
+        intercept = 0
+
+    if n_samples < n_features:
+        warnings.filterwarnings("ignore", category=scipy.linalg.LinAlgWarning)
```
I delete this filterwarning as it is not needed for now.

lorentzenchr committed:
CLN remove unnecessary filterwarnings

lorentzenchr committed:
CLN address review comments

lorentzenchr committed:
Merge branch 'glm_tests' of https://github.com/lorentzenchr/scikit-learn into glm_tests

ogrisel commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,476 @@ def regression_data():
+    assert np.linalg.matrix_rank(X) <= min(n_samples, n_features)
+
+    if fit_intercept and n_samples < n_features:
+        warnings.filterwarnings("ignore", category=ConvergenceWarning)
```
We should not set `filterwarnings` outside of a `warnings.catch_warnings` context manager to avoid having this test introducing a global side effect.

Also, is there a valid reason to expect this warning to show up in this case? If not we should at least add an inline comment with `XXX: investigate why LBFGS fails to properly convege in this case`.

lorentzenchr replied:
What do you propose instead of `warnings.filterwarnings`? Using a pytest decorator is not that granular. It really only happens in this case (`if` condition).

ogrisel replied:
Here is a suggestion: https://github.com/scikit-learn/scikit-learn/pull/23619/files#r900054999

ogrisel commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,476 @@ def regression_data():
+            # assert norm_model_pen == pytest.approx(norm_solution)
+            # assert model_pen.intercept_ == pytest.approx(intercept)
+            # assert_allclose(model_pen.coef_, coef, rtol=rtol)
+        else:
+            assert model.intercept_ == pytest.approx(intercept)
+            assert_allclose(model.coef_, coef, rtol=rtol)
```
```suggestion
        else:
            # When `fit_intercept=False`, LBFGS naturally converge to the minimum
            # norm solution on this problem.
            # XXX: do we have any theoretical guarantees why this should be the
            # case?
            assert model.intercept_ == pytest.approx(intercept)
            assert_allclose(model.coef_, coef, rtol=rtol)
```

ogrisel committed:
Trigger [all random seeds] on the following tests:
test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

ogrisel commented:
@lorentzenchr feel free to re-push commit e16d04e1ee16e3a26585d4c2972ebe0332525742 whenever you want to all the all the random seeds for the tests introduced or updated in this PR.

lorentzenchr committed:
CLN add comment for lbfgs ftol=64 * machine precision

lorentzenchr committed:
Merge branch 'glm_tests' of https://github.com/lorentzenchr/scikit-learn into glm_tests

lorentzenchr committed:
CLN XXX code comment

lorentzenchr committed:
Trigger [all random seeds] on the following tests:

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

lorentzenchr committed:
CLN link issue and remove code snippet in comment

lorentzenchr committed:
Trigger [all random seeds] on the following tests:

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

ogrisel commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,472 @@ def regression_data():
+        intercept = 0
+        X = 0.5 * np.concatenate((X, X), axis=1)
+    assert np.linalg.matrix_rank(X) <= min(n_samples, n_features)
+
+    if fit_intercept and n_samples < n_features:
+        warnings.filterwarnings("ignore", category=ConvergenceWarning)
+    model.fit(X, y)
```
```suggestion

    with warnings.catch_warnings():
        if fit_intercept and n_samples < n_features:
            # XXX: investigate if the lack of convergence in this case should
            # be considered a bug or not.
            warnings.filterwarnings("ignore", category=ConvergenceWarning)
        model.fit(X, y)
```

lorentzenchr committed:
CLN add catch_warnings

ogrisel commented:
Thinking about this test harness, we could have an extended fixture that would actually leverage the distribution family to sample a random `y_train` from the true model.

- This could be useful to check the compatibility of the range of the inverse link function with the natural domain of the distribution family. But maybe this is a bit redundant with the existing tests that do this checks more directly (e.g. in `sklearn/_loss/tests/test_glm_distributions.py`).

- It could also be useful to check the ability of the models to converge to the true distribution instead of just estimating a good conditional expectile. This could maybe be assessed by computing the full loglikelihood of the data under various models of the EDF (instead of just the deviance) and then check that model with the right choice of the family has the lowest negative loglikelihood. I remember that we had attempted to test something like this with poisson regression and gamma regression but had unexpected results. I cannot find the test any more. Was it with decision tree based models maybe? I am not sure if this is because we where looking at the deviance instead of the loglikelihood.

lorentzenchr commented:
> compatibility of the range of the inverse link function

We already check for valid range of `y` in: https://github.com/scikit-learn/scikit-learn/blob/8c82c7bf326dfe7ad6570e2929dde872045e1a53/sklearn/linear_model/_glm/glm.py#L238

> check the ability of the models to converge (in the large sample limit) to the true distribution 

I think we do not need*** such a large sample check (note: no plural "s" in "sample" 🤣). It would merely check the convergence of the drawn random sample to the true distribution, but it would not tell us whether our GLMs fit correctly. A correctly specified GLM should converge to the true coefficients, not matter which distribution the data (as long as finite 2nd moment). The tests here do this even for finite sample sizes and are therefore even more strict for the GLMs.

***"not need" does not mean "refuse".

BTW, expectiles are like quantiles more general. GLMs only estimate the expected value, i.e. expectile at level 0.5 (like median).

lorentzenchr commented:
If I were to review myself, this LGTM:wink:

ogrisel commented:
Indeed, that would just be a complex way to check the consistency of a nll expression and a sampling procedure of the same distribution. But we do not have the nll expression in the current code so there is no point.

ogrisel reviewed (approved):
I am still +1 :)

ogrisel commented:
Can you please just repush the `[all random seeds]` commit?

lorentzenchr committed:
Trigger [all random seeds] on the following tests:

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

lorentzenchr commented:
> Can you please just repush the [all random seeds] commit?

Locally, all is green with `KLEARN_TESTS_GLOBAL_RANDOM_SEED="all"`.

lorentzenchr commented:
I do not understand the CI failure, e.g. in https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=43429&view=logs&j=dde5042c-7464-5d47-9507-31bdd2ee0a3a&t=4bd2dad8-62b3-5bf9-08a5-a9880c530c94:

<details>

```
Starting: Test Library
==============================================================================
Task         : Command line
Description  : Run a command line script using Bash on Linux and macOS and cmd.exe on Windows
Version      : 2.201.1
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/command-line
==============================================================================
Generating script.
Script contents:
exec build_tools/azure/test_script.sh
========================== Starting Command Output ===========================
/usr/bin/bash --noprofile --norc /home/vsts/work/_temp/92bf75cb-86a5-49eb-a0b0-fc037f3afc71.sh
Number of cores: 2

System:
    python: 3.10.4 | packaged by conda-forge | (main, Mar 24 2022, 17:38:57) [GCC 10.3.0]
executable: /usr/share/miniconda/envs/testvenv/bin/python
   machine: Linux-5.13.0-1029-azure-x86_64-with-glibc2.31

Python dependencies:
      sklearn: 1.2.dev0
          pip: None
   setuptools: 62.3.2
        numpy: 1.22.4
        scipy: 1.8.1
       Cython: 0.29.30
       pandas: 1.4.2
   matplotlib: 3.5.2
       joblib: 1.1.0
threadpoolctl: 3.1.0

Built with OpenMP: True

threadpoolctl info:
       user_api: blas
   internal_api: mkl
         prefix: libmkl_rt
       filepath: /usr/share/miniconda/envs/testvenv/lib/libmkl_rt.so.2
        version: 2022.0-Product
threading_layer: intel
    num_threads: 2

       user_api: openmp
   internal_api: openmp
         prefix: libomp
       filepath: /usr/share/miniconda/envs/testvenv/lib/libomp.so
        version: None
    num_threads: 2
# packages in environment at /usr/share/miniconda/envs/testvenv:
#
... [OUTPUT OMITTED]
zstd                      1.5.2                h8a70e8d_1    conda-forge
+ eval 'python -m pytest --showlocals --durations=20 --junitxml=test-data.xml --cov-config='\''/home/vsts/work/1/s/.coveragerc'\'' --cov sklearn --cov-report= -n2 -ra -k '\''on the following tests: or  or test_glm_regression or test_glm_regression_hstacked_X or test_glm_regression_vstacked_X or test_glm_regression_unpenalized or test_glm_regression_unpenalized_hstacked_X or test_glm_regression_unpenalized_vstacked_X or test_warm_start'\'' --pyargs sklearn'
++ python -m pytest --showlocals --durations=20 --junitxml=test-data.xml --cov-config=/home/vsts/work/1/s/.coveragerc --cov sklearn --cov-report= -n2 -ra -k 'on the following tests: or  or test_glm_regression or test_glm_regression_hstacked_X or test_glm_regression_vstacked_X or test_glm_regression_unpenalized or test_glm_regression_unpenalized_hstacked_X or test_glm_regression_unpenalized_vstacked_X or test_warm_start' --pyargs sklearn
============================= test session starts ==============================
platform linux -- Python 3.10.4, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/vsts/work/tmp_folder, configfile: setup.cfg, testpaths: sklearn
plugins: xdist-2.5.0, forked-1.4.0, cov-3.0.0
gw0 I / gw1 I
gw0 [114068] / gw1 [114068]

INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/_pytest/main.py", line 268, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/_pytest/main.py", line 322, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/pluggy/_hooks.py", line 265, in __call__
INTERNALERROR>     return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/pluggy/_manager.py", line 80, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/pluggy/_callers.py", line 60, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/pluggy/_result.py", line 60, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/pluggy/_callers.py", line 39, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/xdist/dsession.py", line 117, in pytest_runtestloop
INTERNALERROR>     self.loop_once()
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/xdist/dsession.py", line 140, in loop_once
INTERNALERROR>     call(**kwargs)
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/xdist/dsession.py", line 262, in worker_collectionfinish
INTERNALERROR>     self.sched.schedule()
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/xdist/scheduler/load.py", line 257, in schedule
INTERNALERROR>     self._send_tests(next(nodes), 1)
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/xdist/scheduler/load.py", line 269, in _send_tests
INTERNALERROR>     node.send_runtest_some(tests_per_node)
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/xdist/workermanage.py", line 284, in send_runtest_some
INTERNALERROR>     self.sendcommand("runtests", indices=indices)
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/xdist/workermanage.py", line 300, in sendcommand
INTERNALERROR>     self.channel.send((name, kwargs))
INTERNALERROR>   File "/usr/share/miniconda/envs/testvenv/lib/python3.10/site-packages/execnet/gateway_base.py", line 728, in send
INTERNALERROR>     raise IOError("cannot send to {!r}".format(self))
INTERNALERROR> OSError: cannot send to <Channel id=1 closed>

======================= 1 skipped, 74 warnings in 54.01s =======================
##[error]Bash exited with code '3'.
Finishing: Test Library
```

</details>

jjerphan commented:
> I do not understand the CI failure, e.g. in [dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=43429&view=logs&j=dde5042c-7464-5d47-9507-31bdd2ee0a3a&t=4bd2dad8-62b3-5bf9-08a5-a9880c530c94](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=43429&view=logs&j=dde5042c-7464-5d47-9507-31bdd2ee0a3a&t=4bd2dad8-62b3-5bf9-08a5-a9880c530c94):

This looks like IPC failures for pytest-xdist, which are unrelated to this PR. What if you merge `main` and re-trigger the CI?

lorentzenchr committed:
Merge branch 'main' into glm_tests

lorentzenchr committed:
Trigger [all random seeds] on the following tests:

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

jjerphan commented:
At least one tail fails on my machine using:

```
SKLEARN_TESTS_GLOBAL_RANDOM_SEED="all" pytest -x -Werror sklearn/linear_model/_glm/tests/test_glm.py
```


<details>
<summary>Details</summary>

```
========================================================================================================================== FAILURES ==========================================================================================================================
______________________________________________________________________________________ test_glm_regression_vstacked_X[long-TweedieRegressor(power=1.5)-17-False-lbfgs] _______________________________________________________________________________________

solver = 'lbfgs', fit_intercept = False
glm_dataset = (TweedieRegressor(power=1.5), array([[-0.14530813, -0.08971888,  0.37246171,  1.        ],
       [ 0.5643881 ,  0.040...2069159]), array([-0.66332878,  0.29030583, -0.6695594 ,  3.32413548]), array([-1.79552203,  0.85862451, -2.92388429]))

    @pytest.mark.parametrize("solver", SOLVERS)
    @pytest.mark.parametrize("fit_intercept", [True, False])
    def test_glm_regression_vstacked_X(solver, fit_intercept, glm_dataset):
        """Test that GLM converges for all solvers to correct solution on vstacked data.

        We work with a simple constructed data set with known solution.
        Fit on [X] with alpha is the same as fit on [X], [y]
                                                    [X], [y] with 1 * alpha.
        It is the same alpha as the average loss stays the same.
        For wide X, [X', X'] is a singular matrix.
        """
        model, X, y, _, coef_with_intercept, coef_without_intercept = glm_dataset
        n_samples, n_features = X.shape
        alpha = 1.0  # because glm_dataset uses this.
        params = dict(
            alpha=alpha,
            fit_intercept=fit_intercept,
            # solver=solver,  # only lbfgs available
            tol=1e-12,
            max_iter=1000,
        )

        model = clone(model).set_params(**params)
        X = X[:, :-1]  # remove intercept
        X = np.concatenate((X, X), axis=0)
        assert np.linalg.matrix_rank(X) <= min(n_samples, n_features)
        y = np.r_[y, y]
        if fit_intercept:
            coef = coef_with_intercept
            intercept = coef[-1]
            coef = coef[:-1]
        else:
            coef = coef_without_intercept
            intercept = 0
>       model.fit(X, y)

sklearn/linear_model/_glm/tests/test_glm.py:332:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
sklearn/linear_model/_glm/glm.py:298: in fit
    self.n_iter_ = _check_optimize_result("lbfgs", opt_res)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

solver = 'lbfgs'
result =       fun: 51.5511742396126
 hess_inv: <3x3 LbfgsInvHessProduct with dtype=float64>
      jac: array([-2.36830267e-09,...: 64
      nit: 9
     njev: 64
   status: 2
  success: False
        x: array([-1.79552203,  0.85862451, -2.92388432])
max_iter = None, extra_warning_msg = None

    def _check_optimize_result(solver, result, max_iter=None, extra_warning_msg=None):
        """Check the OptimizeResult for successful convergence

        Parameters
        ----------
        solver : str
           Solver name. Currently only `lbfgs` is supported.

        result : OptimizeResult
           Result of the scipy.optimize.minimize function.

        max_iter : int, default=None
           Expected maximum number of iterations.

        extra_warning_msg : str, default=None
            Extra warning message.

        Returns
        -------
        n_iter : int
           Number of iterations.
        """
        # handle both scipy and scikit-learn solver names
        if solver == "lbfgs":
            if result.status != 0:
                try:
                    # The message is already decoded in scipy>=1.6.0
                    result_message = result.message.decode("latin1")
                except AttributeError:
                    result_message = result.message
                warning_msg = (
                    "{} failed to converge (status={}):\n{}.\n\n"
                    "Increase the number of iterations (max_iter) "
                    "or scale the data as shown in:\n"
                    "    https://scikit-learn.org/stable/modules/"
                    "preprocessing.html"
                ).format(solver, result.status, result_message)
                if extra_warning_msg is not None:
                    warning_msg += "\n" + extra_warning_msg
>               warnings.warn(warning_msg, ConvergenceWarning, stacklevel=2)
E               sklearn.exceptions.ConvergenceWarning: lbfgs failed to converge (status=2):
E               ABNORMAL_TERMINATION_IN_LNSRCH.
E
E               Increase the number of iterations (max_iter) or scale the data as shown in:
E                   https://scikit-learn.org/stable/modules/preprocessing.html

sklearn/utils/optimize.py:256: ConvergenceWarning
```

</details>

<details>
<summary>Versions</summary>

```
In [1]: from sklearn import show_versions

In [2]: show_versions()

System:
    python: 3.10.2 | packaged by conda-forge | (main, Mar  8 2022, 15:52:01) [GCC 9.4.0]
executable: /home/jjerphan/.local/share/miniconda3/envs/sk/bin/python
   machine: Linux-5.16.14-200.fc35.x86_64-x86_64-with-glibc2.34

Python dependencies:
      sklearn: 1.2.dev0
          pip: 22.0.4
   setuptools: 60.9.3
        numpy: 1.22.3
        scipy: 1.8.0
       Cython: 0.29.30
       pandas: 1.4.1
   matplotlib: 3.5.1
       joblib: 1.1.0
threadpoolctl: 3.1.0

Built with OpenMP: True

threadpoolctl info:
       user_api: blas
   internal_api: openblas
         prefix: libopenblas
       filepath: /home/jjerphan/.local/share/miniconda3/envs/sk/lib/libopenblasp-r0.3.18.so
        version: 0.3.18
threading_layer: pthreads
   architecture: SkylakeX
    num_threads: 8

       user_api: openmp
   internal_api: openmp
         prefix: libgomp
       filepath: /home/jjerphan/.local/share/miniconda3/envs/sk/lib/libgomp.so.1.0.0
        version: None
    num_threads: 8
```


</details>

lorentzenchr committed:
[all random seeds]

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

lorentzenchr commented:
I could make the CI work in https://github.com/scikit-learn/scikit-learn/pull/23619/commits/3458c397085585a397117ea210aa3dddbaef6a13. It passes, but on every setting/system several warnings are triggered. Locally, I do not get any error (and I ran it with `-Werror`).

lorentzenchr committed:
Trigger with -Werror [all random seeds]

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

lorentzenchr committed:
ENH increase maxls to 50

lorentzenchr committed:
[all random seeds]

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

lorentzenchr committed:
Revert "Trigger with -Werror [all random seeds]"

This reverts commit 99f4cf99ca41b4ad2bdad537ad60f936970e3a88.

lorentzenchr commented:
In https://github.com/scikit-learn/scikit-learn/pull/23619/commits/79ec862e0bc0d8c6c5bde8b0253c1ebde4355ae0 I increased the maximum number of linesearch steps in LBFGS from 40 to 50. This eliminated most convergence warnings:
```
sklearn.exceptions.ConvergenceWarning: lbfgs failed to converge (status=2):
ABNORMAL_TERMINATION_IN_LNSRCH.
               
Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
```
The following plattforms still have warnings:
- Linux ubuntu_atlas 
  <details>

  ```
  System:
      python: 3.8.10 (default, Mar 15 2022, 12:22:08)  [GCC 9.4.0]
  executable: /home/vsts/work/1/s/testvenv/bin/python
     machine: Linux-5.13.0-1029-azure-x86_64-with-glibc2.29

  Python dependencies:
        sklearn: 1.2.dev0
            pip: 20.0.2
     setuptools: 44.0.0
          numpy: 1.17.4
          scipy: 1.3.3
         Cython: 0.29.30
         pandas: None
     matplotlib: 3.1.2
         joblib: 1.0.0
  threadpoolctl: 2.0.0
  ```

  </details>

- Windows py38_pip_openblas_32bit
  <details>

  ```
  System:
      python: 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:34:34) [MSC v.1928 32 bit (Intel)]
  executable: D:\a\1\s\testvenv\Scripts\python.exe
     machine: Windows-10-10.0.20348-SP0

  Python dependencies:
        sklearn: 1.2.dev0
            pip: 21.1.1
     setuptools: 56.0.0
          numpy: 1.22.4
          scipy: 1.8.1
         Cython: 0.29.30
         pandas: None
     matplotlib: None
         joblib: 1.1.0
  threadpoolctl: 3.1.0
  ```
    </details>

**Summary**: I think this is good to go.

lorentzenchr committed:
trigger normal CI

ogrisel commented:
@jeremiedbb we have a problem with the "[all random seeds]" CI config as can be seen in [4fd1d9b](https://github.com/scikit-learn/scikit-learn/pull/23619/commits/4fd1d9bfa3e4c9bcd3b97f0c9ea748f2bec04934) which triggers:

```
INTERNALERROR> OSError: cannot send to <Channel id=1 closed>
```

it might the problem described in this recent-ish comment on a very old gist from 2015:

https://gist.github.com/khamidou/6b7e8702f8fc0e8dd251?permalink_comment_id=4080170

>  I encountered this issue in the context of working with Python and non-Python test files. When using a metafunc to parametrize values in Python tests for a specific test run, this error is thrown due to the presence of non-Python test files.
> The workaround is to exclude the non-Python test files by using pytest_ignore_collect.

I am not sure what the non-Python test files could be in our case. Let me try to run "[all random seeds]" without pytest-xdist on this PR.


The original problem from 2015 related to execnet 1.2. However nowadays we use execnet 1.9 so the original workaround to pin execnet to 1.1 is probably out of date.

ogrisel committed:
DEBUG disable pytest-xdist and run [all random seeds]

test_glm_regression
test_glm_regression_hstacked_X
test_glm_regression_vstacked_X
test_glm_regression_unpenalized
test_glm_regression_unpenalized_hstacked_X
test_glm_regression_unpenalized_vstacked_X
test_warm_start

ogrisel commented:
The INTERNALERROR we observed with pytest-xdist seems to go away when running the "all random seeds" tests sequentially (but then everything is very slow). Maybe the problem is related to the very large number of tests we generate when we enable "all random seeds"?

ogrisel committed:
Reenable pytest-xdist on CI

ogrisel committed:
Merge branch 'main' into glm_tests

lorentzenchr committed:
review comment from other PR ;-)

lorentzenchr committed:
Merge branch 'glm_tests' of https://github.com/lorentzenchr/scikit-learn into glm_tests

lorentzenchr commented:
All green. Maybe @jjerphan might have a short look and hit the merge button? This is 99% tests and the one actual change is in the options of lbfgs which lets it pass the tests.

jjerphan reviewed (approved):
LGTM up to minor suggestions. Thank you, @lorentzenchr.

Looking forward to further work on the suite of solvers.

lorentzenchr commented on sklearn/linear_model/_glm/tests/test_glm.py:
```diff
@@ -29,6 +64,475 @@ def regression_data():
+            warnings.filterwarnings("ignore", category=ConvergenceWarning)
+        model.fit(X, y)
+
+    if fit_intercept and n_samples < n_features:
```
It does not make a difference, but ok.

lorentzenchr committed:
CLN address some review comments

jjerphan commented on doc/whats_new/v1.2.rst:
```diff
@@ -146,6 +152,12 @@ Changelog
+- |Enhancement| :class:`linear_model.GammaRegressor`,
+  :class:`linear_model.PoissonRegressor` and :class:`linear_model.TweedieRegressor`
+  can reach higher precision with the lbfgs solver, in particular when `tol` is set
+  to a tiny value.
```
Oops, I should mentioned it previously.
```suggestion
  to a tiny value. Moreover, `verbose` is now properly propagated to L-BFGS-B.
```

lorentzenchr committed:
CLN add comment on solver parameter

lorentzenchr committed:
DOC align whatsnew entries

ogrisel committed:
Remove xfail from test_glm_regression_hstacked_X [all random seeds]

test_glm_regression_hstacked_X

ogrisel commented:
All green, the xfail was no longer needed. Let's merge.
